docs/dev/operators.md

```diff
@@ -101,13 +101,42 @@ spec:
-Store the fact that operator ingress-operator uses both of those images in a metadata file alongside the manifests
-Bundle up your manifests and the metadata file as a docker image and push them to a registry
+1. Verify that the tags `ingress-operator` and `haproxy-router` exist from the release / CI tooling (in the image stream `openshift/origin-v4.0` on api.ci).  If they don’t exist, you’ll get a build error.
+2. Do a find and replace in your manifests (effectively a `sed`) that replaces `quay.io/openshift/origin-haproxy-router(:.*|@:.*)` with `registry.svc.ci.openshift.org/openshift/origin-v4.0@sha256:<latest SHA for :haproxy-router>`
+3. Store the fact that operator ingress-operator uses both of those images in a metadata file alongside the manifests.
+4. Bundle up your manifests and the metadata file as a docker image and push them to a registry.
 
 Later on, when someone wants to mirror a particular release, there will be tooling that can take the list of all images used by operators and mirror them to a new repo.
 
 This pattern tries to balance between having the manifests in your source repo be able to deploy your latest upstream code *and* allowing us to get a full listing of all images used by various operators.
+
+## Dynamic objects
+
+Some objects are not static.
+For example, an operator configuration could depend on configuration passed to the installer (cluster name, cluster domain, service CIDR, etc.).
+An operator should auto-discover the dependencies and build the dynamic objects when it first comes up.
+For example:
+
+1. An operator provides `/manifests/deployment.yaml` with a static deployment.
+2. The cluster-version operator processes the operator's [`/manifests`](#what-do-i-put-in-manifests) and pushes the deployment into the cluster.
+3. The cluster schedules the operator for a particular node.
+4. That node launches the operator pod and its constituent containers.
+5. The operator container tries to pull its configuration.
+    * If found, the operator uses that configuration.
+    * If not found, the operator:
+        1. Pulls the dependency resources (e.g. the `cluster-config-v1` config-map from the `kube-system` namespace).
+            Documentation for connecting to the Kubernetes API from pod containers is [here][API-from-pod].
+            Operators written in Go can use [`InClusterConfig`][InClusterConfig] to configure their client.
+        2. Generates the dynamic configuration based on the dependencies.
+        3. Pushes the generated configuration into the cluster, where future operator runs and other consumers will be able to find it.
+
+For an example, see [the machine-API operator's `getOperatorConfig`][getOperatorConfig], although they don't need the "push the generated configuration into the cluster" step.
```

Comment:
Maybe we should inline the example in a code block here.

Reply:
> Maybe we should inline the example in a code block here.

How much of the code?  e24bc29450 currently links to the main portion of this, and I mention `InClusterConfig` further up but don't link to [the operator's `InClusterConfig` call][1] or [client construction][2].  Putting that all together into one example would be something like:

```go
include (
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
  "k8s.io/client-go/rest"
)

func example() {
  config, err := rest.InClusterConfig() // FIXME: error handling
  client := kubernetes.NewForConfigOrDie(rest.AddUserAgent(cb.config, "your-operator")).CoreV1().ConfigMaps("kube-system")
  for {
    config, err := getConfig(client) // FIXME: error handling
    // FIXME: do whatever your operator does
  }
}

func getConfig(client corev1.ConfigMapInterface) (FIXMESomeType, error) {
  clusterConfig, err := client.Get("cluster-config-v1", metav1.GetOptions{}) // FIXME: error handling
  return yourConfigGenerator(clusterConfig)
}
```

That's a lot of FIXMEs for things I've skipped, and it's still fairly involved.  Is it particularly helpful when we're already linking folks to the not-much-more-complicated real-world machine-API operator implementation? 

[1]: https://github.com/openshift/machine-api-operator/blob/v0.1.0/cmd/common/client_builder.go#L56
[2]: https://github.com/openshift/machine-api-operator/blob/v0.1.0/cmd/common/client_builder.go#L32